Add listing of users (backend)

EvanAntoni opened the pull request:
As agreed. Small pull requests.
Check it out, please.

EvanAntoni committed:
Add listing of users

RazSKg commented on easyrest/models/validator.py:
```diff
@@ -29,3 +31,98 @@ def validation(schema, data):
+            'Admin': []
+        },
+        'Waiter': {
+            'Client': ['create'],
```
Is this normal, when Waiter can create client? Or by "create" you mean other action (besides client creation/modification)?

EvanAntoni replied:
> Is this normal, when Waiter can create client?

I think yes. Create client == Sign Up. Any role can create a user with the role "Client".
I tried to make this access list flexible, so I did not focus on such things. I expected that I would customize actions as further developed. So this is not the final version.

RazSKg commented on easyrest/models/validator.py:
```diff
@@ -29,3 +31,98 @@ def validation(schema, data):
+            'Admin': []
+        },
+        'Owner': {
+            'Client': ['create'],
```
The same question as above.

EvanAntoni replied:
UP

RazSKg commented on easyrest/models/validator.py:
```diff
@@ -29,3 +31,98 @@ def validation(schema, data):
+            'Client': ['create', 'read', 'delete'],
+            'Waiter': [],
+            'Administrator': [],
+            'Owner': ['read'],
```
Moderator can not only read owner.

EvanAntoni replied:
I will edit the actions as necessary. If you need to add editing you can just add 'edit'. Or you can simply add the action 'toggle_activity' so that the moderator can switch only the status field.

RazSKg commented on easyrest/models/validator.py:
```diff
@@ -29,3 +31,98 @@ def validation(schema, data):
+    except KeyError as ke:
+        log.error('Incorrect role "%s"', ke.message)
+        raise HTTPNotFound('Role not found')
+    return True
```
This function never return False. Only True and Exceptions. 😊

EvanAntoni replied:
Oops. Looks like an unnecessary string.

RazSKg commented on easyrest/models/validator.py:
```diff
@@ -29,3 +31,98 @@ def validation(schema, data):
+    log = logging.getLogger(__name__)
+    if not database.query(database.query(column).filter(column == value).exists()).scalar():
+        log.error('{} not found'.format(value))
+        raise HTTPNotFound('{} not found'.format(value))
```
In this case - is it right to raise NotFound rather than BadRequest?

EvanAntoni replied:
I used HTTPNotFound to immediately send a message to the frontend that the value was not found. Since the function checks the presence of a value in the column, I think it is logical to raise this exception.

EvanAntoni committed:
Modifу checking the existence and access of user roles

Changed the check for the existence of the requested role in the function "get_users_list".
Removed the function "check_for_exist" because it has become unnecessary.
Changed the role rights check in the function "check_action_access".

MaxSukhovetskyi commented on easyrest/models/validator.py:
```diff
@@ -29,3 +30,75 @@ def validation(schema, data):
+            'Admin': []
+        }
+    }
+    if action not in user_access_list[user_role][foreign_role]:
```
You have no need to define roles with empty actions. You can use user_access_list[user_role].get(foreign_role, [])
to get same result with less code

EvanAntoni replied:
But in this case, the visibility of actions is lost + this is not the final version. I plan to make changes to this dictionary in the future. But I will take into account your comment.

MaxSukhovetskyi commented on easyrest/views/user_controller.py:
```diff
@@ -50,3 +55,58 @@ def sign_up(request):
+        current_user = request.token.user
+    except AttributeError as ae:
+        log.error(ae.message)
+        raise HTTPForbidden('Need token for access')
```
Why you do not use auth.restrict_access(user_types=['Client', 'Owner', 'Moderator', 'Admin'....])?

EvanAntoni replied:
You're right. In this case, it is better to use the restrict_access.

EvanAntoni replied:
By the way, access restriction can only be used as a decorator, is not it?
`restrict_access` will not raise an exception if I use it in the function body as `restrict_access(user_types=['Client', 'Owner', 'Moderator', 'Admin'])`, am I right?

For example:
```
from ..auth import restrict_access

def get_users_list(request):
    restrict_access(user_types=['Client', 'Owner', 'Moderator', 'Admin'])
    # There must be an exception HTTPForbidden("Need token for access")
    # if the user is not registered but the script continues to be executed further  
    
    derivable_role_id = int(request.matchdict['role_id'])
    role = request.dbsession.query(UserRole).get(derivable_role_id)
    if role is None:
        raise HTTPNotFound(request.path)
```

MaxSukhovetskyi replied:
Why you cant use it as a decorator above def get_users_list(request):?

EvanAntoni replied:
So that's what I did in this case. See `eee54c5`

EvanAntoni committed:
Replace token verification code with "restriction_access" decorator

EvanAntoni committed:
Add user creation

serge30 commented on easyrest/views/user_controller.py:
```diff
@@ -50,3 +55,58 @@ def sign_up(request):
+    users_list = [user.as_dict(exclude=['role_id', 'password']) for user in
+                  request.dbsession.query(User).filter_by(role_id=derivable_role_id).order_by(User.name).all()]
+
+    return wrap(users_list, success=True, message='Users with role \'{}\''.format(role.name))
```
You may use double quotes to avoid escaping:
```... message="Users with role '{}'".format(...)```

EvanAntoni replied:
Yes, I know about it. But I wanted to stick with common styles. In addition, I wanted to have a response in the format of single quotes like that:
```
{
    "message": "Users with role 'Client'"
}
```

EvanAntoni committed:
Edit docstrings and formatting

EvanAntoni committed:
Add user updation

EvanAntoni committed:
Change json format check and change user add behavior

EvanAntoni committed:
Change docstrings

EvanAntoni committed:
Add user deletion

EvanAntoni committed:
Edit user deletion

Simplify user deletion logic. The deletion method is removed
from the user model and implemented in the controller.

EvanAntoni committed:
Merge branch 'users_list' into users_create

EvanAntoni committed:
Merge pull request #37 from Social-projects-Rivne/users_create

Add user creation (backend)

EvanAntoni committed:
Merge branch 'users_list' into users_update

EvanAntoni committed:
Merge pull request #40 from Social-projects-Rivne/users_update

Users update (backend)

EvanAntoni committed:
Merge branch 'users_list' into users_delete

EvanAntoni committed:
Merge pull request #42 from Social-projects-Rivne/users_delete

Users deletion (backend)

EvanAntoni committed:
Edit email checking in user model update method
